File: backup/cses/Python/Additional Problems I/2422 | Multiplication Table.py
# Submission ID: 16997253
# Problem: Multiplication Table
# Link: https://cses.fi/problemset/task/2422

import sys


def solve():
    n = int(sys.stdin.readline().strip())

    l, r = 1, n * n
    target = (n * n + 1) // 2

    while l < r:
        mid = (l + r) // 2
        cnt = 0

        for i in range(1, n + 1):
            cnt += min(mid // i, n)

        if cnt >= target:
            r = mid
        else:
            l = mid + 1

    print(l)


if __name__ == "__main__":
    solve()

File: backup/cses/C++/Additional Problems I/2422 | Multiplication Table.cpp
#pragma GCC optimize("Ofast")
#include <iostream>
#include <algorithm>
#define AC cin.tie(0)->sync_with_stdio(false);

using namespace std;
typedef long long ll;

void solve () {
    ll n;
    cin >> n;
    ll l = 1, r = n * n;
    ll target = (n * n + 1) >> 1; 
    while (l < r) {
        ll mid = (l + r) >> 1;
        ll cnt = 0;
        for (ll i = 1; i <= n; i++) {
            cnt += min(mid / i, n);
        }
        if (cnt >= target)
        	r = mid;
        else
        	l = mid + 1;
    }
    cout << l << '\n';
}

int main () {
    AC
    solve();
    return 0;
}
